### ISSUE TITLE
[Request] add caption words attention/emphasys for training

### ISSUE THREAD

**Issue by swfsql** [kohya-ss/sd-scripts#312]
Is it possible to implement emphasys changes on words from the captions during training?  
This is the explanation of this feature in [auto](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#attentionemphasis).

Example: "an astronaut riding a horse on the (moon:1.5)"
Thanks in advance!

Edit: I'm seeing that there is a regex related to this in [here](https://github.com/kohya-ss/sd-scripts/blob/main/library/lpw_stable_diffusion.py), but would someone know if this is applied during training? Thanks!

**Comment by AI-Casanova** [comment]
The code from `lpw_stable_diffusion.py` doesn't look too hard to implement, and the negative prompts bits can be dropped. 

I'll give it a shot when I have time, and ping @swfsql in the PR if it works to help do some testing for effectiveness.

**Comment by swfsql** [comment]
Thanks, please let me know if I can be of help with the testing.
I'm somewhat new to both python, AI and SD, and new to LoRa and training and so on, but I was trying to train a LoRa on gif images, and I had no luck so far.. and I also only got 6GB of vram.

But my original thought was that I could try separating the gif files in different "duration" concepts, where I would try to correlate the actual gif duration with the attention of the "duration" token(s) according to some formula. For example, saying `(gif duration:0.5)` could mean the gif tagets 0.5s of duration, and saying `(gif duration:1.5)` could mean the gif targets 5s of duration..

I'm pretty sure this would not work, as I only had failures so far, but anyway I couldn't help but try asking if this feature existed XD But coming back to you, I can try coming with other ways to test it in case you implement it, even if it's not related to the case I was thinking originally!

**Comment by AI-Casanova** [comment]
> gif images

Unfortunately the only format that will load are `IMAGE_EXTENSIONS = [".png", ".jpg", ".jpeg", ".webp", ".bmp", ".PNG", ".JPG", ".JPEG", ".WEBP", ".BMP"]`

The way I see this feature working is to emphasize foreground tokens and de emphasize background tokens.

ie

`(SKS woman:1.1), (wearing a black dress:0.9), (sitting in a restaurant:0.8)`

In this way, we can hopefully learn the important features more strongly than the backgrounds, without ignoring the backgrounds completely.

Best practices for captioning will have to come from experimentation, but this will hopefully give us a new tool to experiment with.

**Comment by AI-Casanova** [comment]
Feel free to test out the PR. Will add Dreambooth and finetuning as soon as I'm able.

**Comment by swfsql** [comment]
Thanks @AI-Casanova , I'm looking into testing this change. I'm using the [linux docker](https://github.com/P2Enjoy/kohya_ss-docker) version of the code, and it seems that it's behind from some file refactors that has happened. I'll try applying the changes locally and see then try to see what differences it gives!

**Comment by AI-Casanova** [comment]
re: loss graphs

Loss is nearly completely determined by the number of training timesteps, which under set seed (deterministic training) track exactly. The loss graph on the PR is actually orange and yellow (thanks TensorBoard!) Tracking exactly, but producing these different results, all else held equal. 
![xyz_grid-0192-RPGv4](https://user-images.githubusercontent.com/54461896/228508511-888f6aab-fc19-4ef3-8182-64593d422b1b.jpg)

I'll be doing more testing myself once work calms down.
